### ISSUE TITLE
Cannot reuse a class with dataclass though works with BaseModel

### ISSUE THREAD

**Issue by tsuga** [samuelcolvin/pydantic#4290]
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.9.1
            pydantic compiled: True
                 install path: /usr/local/lib/python3.7/dist-packages/pydantic
               python version: 3.7.13 (default, Apr 24 2022, 01:04:09)  [GCC 7.5.0]
                     platform: Linux-5.4.188+-x86_64-with-Ubuntu-18.04-bionic
     optional deps. installed: ['typing-extensions']
```

```py
import pydantic
from dataclasses import dataclass

@dataclass
class ClassA():
    foo: str
    bar: float

@dataclass
class ClassB():
    a1: ClassA
    a2: ClassA

pydantic_cls = pydantic.dataclasses.dataclass(ClassB)
schema = pydantic_cls.__pydantic_model__.schema()
```

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
[<ipython-input-5-575c54138c2b>](https://localhost:8080/#) in <module>()
     14 
     15 pydantic_cls = pydantic.dataclasses.dataclass(ClassB)
---> 16 schema = pydantic_cls.__pydantic_model__.schema()

6 frames
/usr/local/lib/python3.7/dist-packages/pydantic/schema.cpython-37m-x86_64-linux-gnu.so in pydantic.schema.field_singleton_schema()

KeyError: <class 'pydantic.dataclasses.ClassA'>
```

However, the same structure with `BaseModel` works fine.
```py
import pydantic
from pydantic import BaseModel

class ClassA(BaseModel):
    foo: str
    bar: float

class ClassB(BaseModel):
    a1: ClassA
    a2: ClassA

pydantic_cls = pydantic.dataclasses.dataclass(ClassB)
schema = pydantic_cls.__pydantic_model__.schema()  # no error :)
```

**Comment by samuelcolvin** [comment]
I wonder if this would be fixed by #2557?

I think honestly at this stage of pydantic V1, we might have to call this a limitation of pydantic.

**Comment by samuelcolvin** [comment]
No answer, so I'll assume this is fixed by #2557.
